A few descriptor improvements to prepare for Taproot support

sipa opened the pull request:
These are a few refactors and non-invasive improvements to the descriptors code to prepare for adding Taproot descriptors.

None of the commits change behavior in any way, except the last one which improves error reporting a bit.

sipa committed:
Make DescriptorImpl support multiple subscripts

So far, no descriptor exists that supports more than one sub-script
descriptor. This will change with taproot, so prepare for this by
changing the m_subdescriptor_arg from a unique_ptr to a vector of
unique_ptr's.

sipa committed:
refactor: move population of out.scripts from ExpandHelper to MakeScripts

There are currently two DescriptorImpl subclasses that rely on the functionality
that ExpandHelper automatically adds subscripts to the output SigningProvider.

Taproot descriptors will have subscripts, but we don't want them in the
SigningProvider's bare script field. To avoid them ending up there, move this
functionality into the specific classes' MakeScripts implementation.

sipa committed:
Account for key cache indices in subexpressions

This has no effect for now, as the only fragments with sub-script
expressions (sh, wsh) only allow one, and don't have key expressions
in them.

A future Taproot descriptor will however violate both, and we want
the keys in different sub-scripts to be assigned non-overlapping
cache indices.

sipa committed:
refactor: split off subscript logic from ToStringHelper

This will allow subclasses to overwrite the serialization of subscript
arguments without needing to reimplement all the rest of the ToString
logic.

sipa committed:
Clean up context dependent checks in descriptor parsing

This changes all context dependent checks in the parser to be
disjunctions of equality checks, rather than also including inequalities.
This makes sure that adding a new context enum in the future won't change
semantics for existing checks.

The error messages are also made a bit more consistent.

sipa committed:
refactor: move uncompressed-permitted logic into ParsePubkey*

This is a preparation for parsing xonly pubkeys, which will complicate
this logic. It's cleaner to put the decision logic close to the public
key parsing itself.

RiccardoMasutti reviewed (approved):
ACK, seems good to me. Waiting for squash and will review again.

sipa commented:
I don't intend to squash. They're all independent improvements.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -576,33 +581,37 @@ class DescriptorImpl : public Descriptor
+        for (const auto& subarg : m_subdescriptor_args) {
+            std::vector<CScript> outscripts;
+            if (!subarg->ExpandHelper(pos, arg, read_cache, outscripts, subprovider, write_cache)) return false;
+            if (outscripts.size() != 1 && m_subdescriptor_args.size() != 1) return false;
```
5a044ac64f3f51dc89a772f512d403c7baa12ae2: I'm a bit confused what this new check is doing

sipa replied:
Gone.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -576,33 +581,37 @@ class DescriptorImpl : public Descriptor
             out.origins.emplace(entry.first.GetID(), std::make_pair<CPubKey, KeyOriginInfo>(CPubKey(entry.first), std::move(entry.second)));
         }
-        if (m_subdescriptor_arg) {
+        if (m_subdescriptor_args.size() == 1) {
```
5a044ac64f3f51dc89a772f512d403c7baa12ae2: why is `m_subdescriptor_args` length 1 a special case?

sipa replied:
Gone.

Sjors reviewed (commented):
Code review 45bc0c04e0f2ea6dcce734652f45cabc6d10d7d5 looks good, but I got confused in one place...

achow101 commented:
Partial ACK. Same questions as Sjors.

sipa committed:
Remove support for subdescriptors expanding to multiple scripts

sipa commented:
@achow101 @Sjors Good that you raise that point.

The original reason why this behavior of MakeScripts-is-invoked-once-per-evaluation existed was to support things like `combo()`, where a single evaluation at a single index results in multiple scriptPubKeys. This was particularly complicated to combine with descriptors that support multiple subdescriptors (like `tr()`) - there isn't really a straightforward way of doing that (cartesian product? requiring the same number of evaluations from each and combining pairwise? ...?). I didn't want to deal with that, so I added the restriction that this once-per-evaluation only existed for descriptors with 1 subdescriptor.

This is all silly, because we don't ever need it. `combo()` is the only descriptor that produces multiple output scripts, and it is only permitted at the top level - so it's not even possible to have it as a subdescriptor. I've just removed support for multiple-outputscripts-from-subscriptor entirely in a separate commit now. Bye.

Sjors commented:
Thanks, I'll review again after rebase.

sipa commented:
Rebased.

Sjors commented:
ACK 05c633e

Conceptually I find it marginally simpler if you reverse the first and second commit, so the first step is to remove the unused loop:

```cpp
        if (m_subdescriptor_arg) {
            const auto& subscript = subscripts[0];
            out.scripts.emplace(CScriptID(subscript), subscript);
            output_scripts = MakeScripts(pubkeys, &subscript, out);
        } else {
            output_scripts = MakeScripts(pubkeys, nullptr, out);
        }
 ```
 
 And then the second commit drops `out.scripts.emplace` and moves that into `MakeScripts` (which is also where `combo` does this).
 
To sanity check my understanding: `key_exp_index` is currently used for multisig descriptors to create a seperate cache for each participating key. With taproot, you might have a root key (`key_exp_index == 0`), one branch with a key ``key_exp_index == 2`, and another branch with a 3 key multisig (`key_exp_index  3,4,5). Is that how it works?

achow101 commented:
ACK 05c633e0efaf300aa3353303f77c0a622647a588

sipa commented:
@Sjors 

> Conceptually I find it marginally simpler if you reverse the first and second commit

Ok, done.

> With taproot, you might have a root key (`key_exp_index == 0`), one branch with a key `key_exp_index == 2`, and another branch with a 3 key multisig (`key_exp_index  3,4,5`). Is that how it works?

I think you skipped the 1, but otherwise: exactly.

achow101 commented:
re-ACK 35cbef1c483ac9905331f8035d32600e105baab4

Only change was commit reordering.

Sjors commented on src/script/descriptor.cpp:
```diff
@@ -480,34 +480,36 @@ class DescriptorImpl : public Descriptor
-    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::unique_ptr<DescriptorImpl> script, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_arg(std::move(script)) {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args() {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::unique_ptr<DescriptorImpl> script, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args(Vector(std::move(script))) {}
+    DescriptorImpl(std::vector<std::unique_ptr<PubkeyProvider>> pubkeys, std::vector<std::unique_ptr<DescriptorImpl>> scripts, const std::string& name) : m_pubkey_args(std::move(pubkeys)), m_name(name), m_subdescriptor_args(std::move(scripts)) {}
```
warning: unused member function 'DescriptorImpl'

sipa replied:
Fixed.

Sjors reviewed (commented):
re-ACK 35cbef1c483ac9905331f8035d32600e105baab4 modulo unused constructor

S3RK reviewed (commented):
Code Review ACK 394e387. Nice and clear improvements!

sipa commented on src/script/descriptor.cpp:
```diff
@@ -577,30 +579,29 @@ class DescriptorImpl : public Descriptor
+            assert(outscripts.size() == 1);
+            subscripts.emplace_back(std::move(outscripts[0]));
         }
+        out = Merge(std::move(out), std::move(subprovider));
```
It doesn't matter now apparently, as Merge takes const lvalue references only. But a more efficient variant could be added that takes advantage of inputs that aren't needed anymore.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -543,12 +554,7 @@ class DescriptorImpl : public Descriptor
-            if (!scriptarg->ToStringHelper(arg, tmp, priv, normalized)) return false;
-            ret += std::move(tmp);
-        }
+        if (!ToStringSubScriptHelper(arg, ret, priv, normalized, pos)) return false;
```
Done.

sipa commented on src/script/descriptor.cpp:
```diff
@@ -621,9 +622,9 @@ class DescriptorImpl : public Descriptor
         }
-        if (m_subdescriptor_arg) {
+        for (const auto& arg : m_subdescriptor_args) {
             FlatSigningProvider subprovider;
```
Done. In fact, I don't need any at all.

sipa commented:
Rebased, and addressed a few comments.

S3RK commented:
reACK 0b188b7
The only changes are addressing review comments above. 1) removing unused constructor 2) removing redundant signing provided instance creation 3) simplifying `ToStringSubScriptHelper` signature

Sjors commented:
re-ACK 0b188b7

achow101 commented:
re-ACK 0b188b751f970027c52729e0c223cc9257669322

Sjors commented:
@instagibbs what do you think?

instagibbs commented:
Seems like 3 ACKs is enough? not familiar enough with this code anymore to give a proper ACK

ysangkok commented on src/script/descriptor.cpp:
```diff
@@ -1045,6 +1066,9 @@ std::unique_ptr<DescriptorImpl> ParseScript(uint32_t key_exp_index, Span<const c
         auto bytes = ParseHex(str);
         return std::make_unique<RawDescriptor>(CScript(bytes.begin(), bytes.end()));
+    } else if (Func("raw", expr)) {
+        error = "Can only have raw() at top level";
```
why is there no new test for this and for `addr()`?
